MaxText/layers/linears.py

```diff
@@ -404,7 +403,14 @@ def unpermute(self, intermediate, sorted_selected_experts, weights):
           reshaped_weights.astype(jnp.float32),
           precision=matmul_precision,
       )
-    return output.reshape(-1, self.config.max_target_length, self.config.emb_dim // tensor_parallelism).astype(self.dtype)
+    updated_batch = int(self.config.per_device_batch_size * jax.device_count() // self.config.ici_fsdp_parallelism)
```

Comment:
Does this cover all cases - what if we use data_parallelism instead of FSDP (or a mix of them) or even fsdp_transpose?

Reply:
Can we get the relevant sizes from the input shape instead of based on the config?

Reply:
@ZhiyuLi-goog made change there. Would you take a look on this?

Reply:
Thank you for the comment @gobbleturk 

> Does this cover all cases - what if we use data_parallelism instead of FSDP (or a mix of them) or even fsdp_transpose?

Updated to cover data and fsdp (or a mix of them). Not yet fsdp_transpose since it is simply not used in MoE models.

> Can we get the relevant sizes from the input shape instead of based on the config?

It seems not easy to have a relevant sizes solution covering both prefill and decoding situations and the prefill length might be variant.  And this is the trick working pretty well in our inferencing experiments.

@RissyRan @mailvijayasingh feel free to chime in if you have new ideas.

Reply:
> Can we get the relevant sizes from the input shape instead of based on the config?

Yeah, this is the difference between training and inference. The batch size will be different in prefill and generate stage (even we specify per_device_batch_size as a fixed value). In megablox, we need extra permute and unpermute operations, and those need to be calculated manually to map back the right shape. 

@gobbleturk what's your suggestion on this? basically we need an indicator to see if current is prefill or generate stage in inference, and then reshape the right shape, based on different batch size.

Reply:
I have updated this to inference output shape based on input tensor.
But I am not fully familiar with this code, and afraid I might made a wrong implementation here.

@ZhiyuLi-goog Can you take a look on this.
Also @gobbleturk Can you review this branch? It would be great to merge this branch before holiday.

cc. @RissyRan

Reply:
I was thinking something like sharded_batch_shape = inputs.shape[0]. I think prefill and generate are two separate jits (unsure) so this sharded_batch_shape just takes exactly the value it needs